notes/languages/go.md

```diff
@@ -755,3 +755,201 @@ There is also the lack of enums, which to be honest annoys me terribly
+that then magically works, but there is no way for you to implement one:
+
+```
+package main
+
+import (
+	"fmt"
+)
+
+type x struct {
+	a int
+}
+
+func eq[T comparable](a, b T) bool {
+	return a == b
+}
+
+func main() {
+	fmt.Println(eq(1, 1))
+	fmt.Println(eq(1, 2))
+
+	fmt.Println(eq(x{1}, x{1}))
+	fmt.Println(eq(x{1}, x{2}))
+}
+```
+
+So this works and is possible, but only thanks to Go introducing more magical
+concepts that you can't implement yourself. Go already used to do that, it has a lot
+of built-ins that are generic and just work nicely. That is an OK design decision
+in a language, but the whole purpose of a language that has a proper parametric
+polymorphic system is that any of that is not needed, if you design a language
+from scratch with that idea in mind the type system should be powerful enough
+that you don't need reserved/magical things that users can't reproduce, you
+just provide useful collections of functions that do those things.
+
+Go didn't started with this premise, so you end up with a mix of the worse of
+both worlds. You have all the complexity a more sophisticated type system brings
+but you don't have all the benefits, because previous design decisions constrains
+you, so you still need the complexity of magical language defined constraints.
+
+Depending on other barriers found, the language may end up with more magical
+constraints, since obviously not everything that should be possible in a parametric
+polymorphic system is actually possible.
+
+
+#### Messed Up Interfaces
+
+This one is less a limitation on the parametric type system and more a
+bad choice, IMHO, on how to define constraints. They made the decision that
+interfaces are going to be used for type constraints. Which seems to make
+sense when you think about constraints that requires a pre-determined
+set of methods to be available on a type, just like an interface.
+
+But things get pretty bad when you need something like defining an interface
+of anything that can be ordered. Again since Go doesn't have first class
+operators there is no way to express that on a interface. So the hack this
+time was to just "solve" it with a list of concrete types, like what is done on
+the [constraints.Ordered](https://cs.opensource.google/go/x/exp/+/a9213eeb:constraints/constraints.go;drc=8176bd3bf02641480aa57900d370148cff889042;l=48).
+
+First lets just stop and think on the clumsy hack that it is to define a
+constraint of comparable things by having to enumerate ALL THE TYPES THAT
+ARE COMPARABLE. Just that already smells like really bad parametric types.
+But on top of that, the decision was made to do that by extending the current
+interface type. So now an interface can be something like this:
+
+```go
+type a interface {
+    string | int
+}
+```
+
+Now, you may be thinking, can I actually use that as an interface ? Like this:
+
+```go
+func f(x a) {}
+```
+
+No you can't. Go just created a whole new brand of interfaces that can't be
```

Comment:
```
Interfaces that are not [basic](https://go.dev/ref/spec#Basic_interfaces) may only be used as type constraints, or as elements of other interfaces used as constraints. They cannot be the types of values or variables, or components of other, non-interface types.
```

It really looks like a lazy design... they just called the old interface "basic" but the other "general interface"? but the general is the one with actual constrained use cases (less usability). The "basic" can be used everywhere, even in type constraints...

Reply:
yeah..I just don't get it ...maybe I'm missing something, but the more I dig the less it makes sense and the hacky feelings intensifies x_x.